asmcomp/x86_proc.ml

```diff
@@ -71,22 +71,22 @@ let string_of_string_literal s =
   Buffer.contents b
 
 let string_of_symbol prefix s =
-  let spec = ref false in
-  for i = 0 to String.length s - 1 do
-    match String.unsafe_get s i with
-    | 'A'..'Z' | 'a'..'z' | '0'..'9' | '_' | '.' -> ()
-    | _ -> spec := true;
-  done;
-  if not !spec then if prefix = "" then s else prefix ^ s
+  let symbol_separator = Compilenv.symbol_separator () in
+  let is_spec_char = function
+    | 'A'..'Z' | 'a'..'z' | '0'..'9' | '_' -> false
+    | c -> c <> symbol_separator
+  in
+  let spec = String.exists is_spec_char s in
+  if not spec then if prefix = "" then s else prefix ^ s
   else
     let b = Buffer.create (String.length s + 10) in
     Buffer.add_string b prefix;
     String.iter
-      (function
-        | ('A'..'Z' | 'a'..'z' | '0'..'9' | '_' | '.') as c ->
-            Buffer.add_char b c
-        | c ->
-            Printf.bprintf b "$%02x" (Char.code c)
+      (fun c ->
+       if is_spec_char c then
+         Printf.bprintf b "$%02x" (Char.code c)
```

Comment:
> Maybe we're OK because qualified identifiers in the back-end are `Module.component` or `Module.Submodule.component` (after packing), and `Module` always starts with a capital letter while `component` as well as hexadecimal escapes above always start with a lowercase letter. Still, this is subtle.

I'll add a comment to warn about that.

Reply:
I thought some more about the risk of conflicts.  If we stick to ASCII for identifiers, it should be OK: `$xx` escapes are used only for symbols like `+` or `^^`, and these can only occur in the `component` part of compound identifiers `Module$Submodule$component`.  So, the first `$` not followed by an uppercase letter is the separator that starts the `component` part; all previous `$` are separators; and all following `$` are escapes.

If identifiers can use Latin-1 letters (like yesterday) or bits of UTF8-encoded Unicode (like in #12664), things are more complicated and I don't have a proof.  But another way to avoid the problem would be to use decimal (`$%03d`) instead of hexadecimal (`$%02x`) for escapes, as a separator cannot be followed by a digit.